### PR TITLE
Update cost estimates by device form

### LINKED ISSUE #232: Kotlin | Print actual cost of the run
Now that Flank Kotlin supports multiple devices & platforms, the cost estimation is in need of updating. Using information from the device catalog we can accurately calculate the cost per run. Physical devices are billed at $5/hr and virtual devices are billed at $1/hr.

* Determine how many minutes is used per device type
  * iOS is always physical
* Calculate aggregate cost

2 hours on physical devices + 1 hour of virtual devices would be $11.

### PR DESCRIPTION
**Pull request by vanwagonet**
Calculate the estimated cost from each execution device based its form
(virtual or physical). This is also summing all devices instead of only looking
at the first.

fix #232

### CODE REVIEW

**Comment by bootstraponline** [comment]
Thanks, I'll review this tomorrow

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/json/SavedMatrix.kt [hunk 1]
```diff
@@ -45,12 +45,17 @@ class SavedMatrix(matrix: TestMatrix) {
...
-            testDurationSeconds = step.testExecutionStep.testTiming.testProcessDuration.seconds
-            runDurationSeconds = step.runDuration.seconds
-            billableMinutes = Billing.billableMinutes(testDurationSeconds)
+        billableVirtualMinutes = 0
+        billablePhysicalMinutes = 0
+        matrix.testExecutions.forEach {
+            val step = GcToolResults.getResults(it.toolResultsStep)
+            val billableMinutes = Billing.billableMinutes(step.testExecutionStep.testTiming.testProcessDuration.seconds)
+            val androidDevice = it.environment.androidDevice
+            if (androidDevice != null && AndroidCatalog.isVirtualDevice(androidDevice.androidModelId)) {
```
Comment (bootstraponline) [inline cloud_testing/kotlin_poc/src/main/kotlin/ftl/json/SavedMatrix.kt]: I think you could do something like this:

> AndroidCatalog.isVirtualDevice(androidDevice)

then the null check / androidModelId would move into the helper method.

File: cloud_testing/kotlin_poc/src/main/kotlin/ftl/util/Billing.kt [hunk 2]
```diff
@@ -36,23 +36,39 @@ object Billing {
...
-        val virtualCost = billableMinutes.multiply(VIRTUAL_COST_PER_MIN).setScale(2, RoundingMode.HALF_UP)
+    private fun estimateCosts(billableVirtualMinutes: BigDecimal, billablePhysicalMinutes: BigDecimal): String {
+        val virtualCost = billableVirtualMinutes.multiply(VIRTUAL_COST_PER_MIN).setScale(2, RoundingMode.HALF_UP)
+        val physicalCost = billablePhysicalMinutes.multiply(PHYSICAL_COST_PER_MIN).setScale(2, RoundingMode.HALF_UP)
+        val totalCost = (virtualCost + physicalCost).setScale(2, RoundingMode.HALF_UP)
 
-        val remainder = billableMinutes.toLong()
-        val hours = TimeUnit.MINUTES.toHours(remainder)
-        val minutes = remainder % 60
+        val virtualTime = prettyTime(billableVirtualMinutes)
```
Comment (bootstraponline) [inline cloud_testing/kotlin_poc/src/main/kotlin/ftl/util/Billing.kt]: I like the clean up into a prettyTime helper!

File: cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt [hunk 3]
```diff
@@ -21,4 +21,13 @@ class AndroidCatalogTest {
         assertEquals(IncompatibleModelVersion, AndroidCatalog.supportedDeviceConfig("NexusLowRes", "21"))
         assertEquals(SupportedDeviceConfig, AndroidCatalog.supportedDeviceConfig("NexusLowRes", "23"))
     }
+
+    @Test
+    fun validateIsVirtualDevice() {
+        if (bitrise) return
```
Comment (bootstraponline) [inline cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: with mocking, are we able to use a fake AndroidCatalog on bitrise? I worry about too many `if (bitrise) return` means we're not running the tests in CI at all.

Comment (vanwagonet) [reply cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: the mock server handles the android catalog now, but I got some other credentials error when I tried to allow these tests to run.

Comment (vanwagonet) [reply cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]: https://app.bitrise.io/build/9ab1863db270de36 was the build that errored.
`The Application Default Credentials are not available.`

Comment (bootstraponline) [reply cloud_testing/kotlin_poc/src/test/kotlin/ftl/android/AndroidCatalogTest.kt]:
```
 init {
        FtlConstants.useMock = true
    }
```

may fix that and activate the mock credential logic

**Pull request merged**